### DISCUSSION TITLE
Help Needed - Unable to Boot My OS from a USB Drive

### DISCUSSION THREAD

**Discussion by crispytoast7** [CosmosOS/Cosmos#2881] [General]
#### Issue
I can't figure out how to boot my OS from a flash drive. I read <a href="https://www.gocosmos.org/docs/booting-your-os-from-a-usb-drive/">this</a> article and changed "VMWare" to "ISO Image" in the properties, but whenever I start it, it gives me the following error:
![image](https://github.com/CosmosOS/Cosmos/assets/109392473/d749b289-df34-45c8-ac40-9bc18a9b837b)

But, if I set the run type to "release" I get a error stating that the release configuration is not supported.
Please help, thank you!

#### Version
I'm using the COSMOS User Kit in C#.

**Comment by AAM1075** [comment]
I solved this problem by using [Ventoy](https://www.ventoy.net/en/index.html). You can easily drag & drop your ISO on the USB and then select it from a menu once you've booted from the USB. Rufus and BalenaEtcher fail because they don't see any filesystems in the compiled ISO at the moment. You can find your ISO in your project's "\bin\Debug\net6.0" folder.

I'd also suggest that you use the devkit as it's much more updated.

**Comment by crispytoast7** [comment]
@MEMESCOEP I cannot find the ISO file. What is a possibility it could be called in the "\bin\Debug\net6.0" folder? There is a folder named “IOS”. Could it be within that?

Does devkit mean more complexity? The OS I am building is entirely command-line based, no GUI.

**Comment by AAM1075** [comment]
> @MEMESCOEP I cannot find the ISO file. What is a possibility it could be called in the "\bin\Debug\net6.0" folder? There is a folder named “IOS”. Could it be within that?
> 
> Does devkit mean more complexity? The OS I am building is entirely command-line based, no GUI.

The ISO file should be called `<Project name>.iso`, you might need to compile first before it'll show up.

The devkit does have some fixes for console things iirc, I just don't remember exactly what they were

**Comment by crispytoast7** [comment]
@MEMESCOEP The problem is, I get an error whenever I compile my project as a ISO image.

**Comment by AAM1075** [comment]
> @MEMESCOEP The problem is, I get an error whenever I compile my project as a ISO image.

I've accidentally left out some important information, sorry. You can set it back to "VMWare", as it should still generate an ISO file. The ISO option is kinda broken in the userkit, but I forgot why. I'll try to find some more information.

You should also disable the cosmos debugger when testing on real hardware because it'll fail to connect and the OS won't do anything else

**Comment by crispytoast7** [comment]
> I've accidentally left out some important information, sorry. You can set it back to "VMWare", as it should still generate an ISO file. The ISO option is kinda broken in the userkit, but I forgot why. I'll try to find some more information.

@MEMESCOEP Will try.

___

> I've accidentally left out some important information, sorry. You can set it back to "VMWare", as it should still generate an ISO file. The ISO option is kinda broken in the userkit, but I forgot why. I'll try to find some more information.

How do I do so?

**Comment by AAM1075** [comment]
> > I've accidentally left out some important information, sorry. You can set it back to "VMWare", as it should still generate an ISO file. The ISO option is kinda broken in the userkit, but I forgot why. I'll try to find some more information.
> 
> @MEMESCOEP Will try.
> 
> ___
> 
> > I've accidentally left out some important information, sorry. You can set it back to "VMWare", as it should still generate an ISO file. The ISO option is kinda broken in the userkit, but I forgot why. I'll try to find some more information.
> 
> How do I do so?

In your project properties, set "ISO Image" back to "VMware"

**Comment by crispytoast7** [comment]
@MEMESCOEP Great! I have loaded it the ISO file onto a USB with Rufus. But, when I try to boot it from the USB on a computer, it gets stuck on the screen saying “Finding PCI…”

**Comment by AAM1075** [comment]
> @MEMESCOEP Great! I have loaded it the ISO file onto a USB with Rufus. But, when I try to boot it from the USB on a computer, it gets stuck on the screen saying “Finding PCI…”

I've had this problem too on certain computers, tbh I have no clue what causes it. Cosmos just doesn't like certain computers ig

**Comment by crispytoast7** [comment]
@MEMESCOEP So I cannot use my OS on my old MacBook? Would switching to devkit help? Is there anyone who would know a fix?

**Comment by AAM1075** [comment]
> @MEMESCOEP So I cannot use my OS on my old MacBook? Would switching to devkit help? Is there anyone who would know a fix?

I'm not sure how compatible Cosmos is with MacBooks atm, but maybe one of the coredevs know

**Comment by crispytoast7** [comment]
Who is a "coredev" that I can contact?

Also, do you know if COSMOS works on Windows PCs, @MEMESCOEP?

**Comment by 9xbt** [comment]
cosmos supports x86 so yes

**Comment by 9xbt** [comment]
also coredevs are literally core developers, join the discord server for more help

**Comment by 9xbt** [comment]
also "i need help" is not bugs, you should move this to the discussions page

**Comment by AAM1075** [comment]
> also "i need help" is not bugs, you should move this to the discussions page

bruh

**Comment by AAM1075** [comment]
> Who is a "coredev" that I can contact?
> 
> Also, do you know if COSMOS works on Windows PCs, @MEMESCOEP?

yeah, as of right now cosmos development works best on Windows computers

**Comment by crispytoast7** [comment]
> yeah, as of right now cosmos development works best on Windows computers

@MEMESCOEP Good to know.

____

> Who is a "coredev" that I can contact?

@MEMESCOEP 👀 ⏫

**Comment by AAM1075** [comment]
> > Who is a "coredev" that I can contact?
> 
> @MEMESCOEP 👀 ⏫ 

Normally, we'd ask the helpers (they have the helper role) in the cosmos discord server

**Comment by crispytoast7** [comment]
@MEMESCOEP I apologize, but I do not have a discord. Is there another way I can contact them?

**Comment by AAM1075** [comment]
> @MEMESCOEP I apologize, but I do not have a discord. Is there another way I can contact them?

You might be able to contact them through GitHub discussions

**Comment by MishaProductions** [comment]
> @MEMESCOEP Great! I have loaded it the ISO file onto a USB with Rufus. But, when I try to boot it from the USB on a computer, it gets stuck on the screen saying “Finding PCI…”

Cosmos has this issue with more modern computers

**Comment by AAM1075** [comment]
> > @MEMESCOEP Great! I have loaded it the ISO file onto a USB with Rufus. But, when I try to boot it from the USB on a computer, it gets stuck on the screen saying “Finding PCI…”
> 
> Cosmos has this issue with more modern computers

Do we know why?

**Comment by crispytoast7** [comment]
> Cosmos has this issue with more modern computers

@MishaProductions Will this ever be fixed in the future, or is their a workaround?

**Comment by 9xbt** [comment]
Cosmos is not meant for using it as your daily driver, it doesn’t support USB, doesn’t support UEFI, is only i386, etc. We’re trying to implement all this stuff but its gonna take years man osdev is not as easy as you think

**Comment by crispytoast7** [comment]
@9xbt Completely understood. So let me get this straight - COSMOS does not support UEFI, only i386?

**Comment by crispytoast7** [comment]
Whoops didnt mean to close this. I’m still wondering if UEFI support will ever come.

**Comment by MishaProductions** [comment]
Probably not until x64 is implemented

**Comment by crispytoast7** [comment]
@MishaProductions Sorry for the late response, but do we have any information on when x64 support will come?

**Comment by MishaProductions** [comment]
UEFI support was added in #2828 . I don't know when x64 support will be added

**Comment by crispytoast7** [comment]
@MishaProductions That’s great! Thanks for the update! I’ll try again to load COSMOS on my MacBook and see what happens!

**Comment by 9xbt** [comment]
@DarkOutcast6 I'm 99% sure that won't work because computers that are UEFI are usually x64 and the fact that if you have a pretty new MacBook the chance that it will have an M1 or M2 ARM CPU instead of an Intel x86_64 CPU

**Comment by crispytoast7** [comment]
My MacBook is from 2012, so I know it is intel-based. I’m pretty sure it runs x86, though I'm not 100% sure.

**Comment by crispytoast7** [comment]
@9xbt Do you have any ideas for [this](https://github.com/CosmosOS/Cosmos/issues/2878#issuecomment-1874734135) issue?

**Comment by 9xbt** [comment]
> My MacBook is from 2012, so I know it is intel-based. I’m pretty sure it runs x86, though I'm not 100% sure.

It should run just fins

**Comment by 9xbt** [comment]
> @9xbt Do you have any ideas for this issue?

No, not really, but fyi these drivers were probably written before UEFI support and stuff can be different, which may explain why it won't work

**Comment by zarlo** [comment]
> @DarkOutcast6 I'm 99% sure that won't work because computers that are UEFI are usually x64 and the fact that if you have a pretty new MacBook the chance that it will have an M1 or M2 ARM CPU instead of an Intel x86_64 CPU

cosmos will run fine on any x86 cpu including x86_64/amd64 

and with resent changes it should work on UEFI some things will not work like console but

**Comment by crispytoast7** [comment]
@9xbt @zarlo But even though I have tried using the canvas, nothing will appear, no matter what I try. Why is this? Is it a bug?

**Comment by zarlo** [comment]
what canvas back end are you using?

**Comment by crispytoast7** [comment]
@zarlo Wdym? I'm using the FullscreenCanvas, if that's what you mean. It works fine in VMware.

**Comment by zarlo** [comment]
okay try using

```cs
new VBECanvas()
```
if that does not work then try
```cs
new VGACanvas();
```

this will help work out if there is an issue in the logic used by FullScreenCanvas

**Reply by crispytoast7** [reply]
@zarlo Trying it out now. Crossing my fingers!

**Reply by crispytoast7** [reply]
@zarlo VBECanvas doens't work. trying VGACanvas now.

**Comment by crispytoast7** [comment]
@zarlo Tried using `VGACanvas`, but it didn't even work in VMware, and `VBECanvas` worked in VMware, just not on the real device. What should I do?

**Reply by zarlo** [reply]
dont forget you need to call canvas.Display() for it to show anything with some of the backends

**Reply by crispytoast7** [reply]
@zarlo Already did that in my code.

**Reply by zarlo** [reply]
i guess for now you will have to run it in a VM

**Reply by crispytoast7** [reply]
@zarlo @9xbt There's absolutely no way to run it on my physical hardware? Any ideas that could possibly fix it?

**Comment by crispytoast7** [comment]
Hello. I am sorry if this is annoying, but I’d really like to know if there is any way I can use a workaround to run my OS on my MacBook. If anyone has anything to provide that may be of use to me, please do.

**Reply by 9xbt** [reply]
Tim Cook won't like you doing that and even if you were actually able to boot it you wouldn't be able to use the keyboard or mouse because Cosmos doesn't support USB, plus I think Cosmos only outputs VGA but I'm unsure about that. Anyways I don't think it's gonna be possible to run Cosmos on MacBooks (yet).

**Reply by 9xbt** [reply]
I'm sorry about that but I can't try to actually do it myself since I don't own a MacBook

**Reply by 9xbt** [reply]
\:(

**Reply by crispytoast7** [reply]
@9xbt I guess it’s fine, if it really won’t work. Sorry for all the trouble.
